homeassistant/components/crownstone/listeners.py

```diff
@@ -0,0 +1,147 @@
+
+from homeassistant.core import Event, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_send, dispatcher_send
+
+from .const import (
+    DOMAIN,
+    SIG_CROWNSTONE_STATE_UPDATE,
+    SIG_UART_STATE_CHANGE,
+    SSE_LISTENERS,
+    UART_LISTENERS,
+)
+
+if TYPE_CHECKING:
+    from .entry_manager import CrownstoneEntryManager
+
+
+@callback
+def async_update_crwn_state_sse(
+    manager: CrownstoneEntryManager, ha_event: Event
+) -> None:
+    """Update the state of a Crownstone when switched externally."""
+    switch_event = SwitchStateUpdateEvent(ha_event.data)
+    try:
+        updated_crownstone = manager.cloud.get_crownstone_by_id(switch_event.cloud_id)
+    except KeyError:
+        return
+
+    # only update on change.
+    if updated_crownstone.state != switch_event.switch_state:
+        updated_crownstone.state = switch_event.switch_state
+        async_dispatcher_send(manager.hass, SIG_CROWNSTONE_STATE_UPDATE)
+
+
+@callback
+def async_update_crwn_ability(manager: CrownstoneEntryManager, ha_event: Event) -> None:
+    """Update the ability information of a Crownstone."""
+    ability_event = AbilityChangeEvent(ha_event.data)
+    try:
+        updated_crownstone = manager.cloud.get_crownstone_by_id(ability_event.cloud_id)
+    except KeyError:
+        return
+
+    ability_type = ability_event.ability_type
+    ability_enabled = ability_event.ability_enabled
+    # only update on a change in state
+    if updated_crownstone.abilities[ability_type].is_enabled == ability_enabled:
+        return
+
+    # write the change to the crownstone entity.
+    updated_crownstone.abilities[ability_type].is_enabled = ability_enabled
+
+    if ability_event.sub_type == EVENT_ABILITY_CHANGE_DIMMING:
+        # reload the config entry because dimming is part of supported features
+        manager.hass.async_create_task(
+            manager.hass.config_entries.async_reload(manager.config_entry.entry_id)
+        )
+    else:
+        async_dispatcher_send(manager.hass, SIG_CROWNSTONE_STATE_UPDATE)
+
+
+def update_uart_state(manager: CrownstoneEntryManager, _: bool | None) -> None:
+    """Update the uart ready state for entities that use USB."""
+    # update availability of power usage entities.
+    dispatcher_send(manager.hass, SIG_UART_STATE_CHANGE)
+
+
+def update_crwn_state_uart(
+    manager: CrownstoneEntryManager, data: AdvExternalCrownstoneState
+) -> None:
+    """Update the state of a Crownstone when switched externally."""
+    if data.type != AdvType.EXTERNAL_STATE:
+        return
+    try:
+        updated_crownstone = manager.cloud.get_crownstone_by_uid(
+            data.crownstoneId, manager.usb_sphere_id
+        )
+    except KeyError:
+        return
+
+    if data.switchState is None:
+        return
+    # update on change
+    updated_state = cast(SwitchState, data.switchState)
+    if updated_crownstone.state != updated_state.intensity:
+        updated_crownstone.state = updated_state.intensity
```

Comment:
Why doesn't the library handle its own state update?

Reply:
Not entirely sure what you mean here, the `crownstone-uart` library isn't meant to store any states or user data, so it updates the data stored in the data models of the `crownstone-cloud` library.

Reply:
I'm thinking this state update should be hidden from the library user since it's data coming from a device library and used by a device library. It's a bit weird that the library user needs to set up this connection.

Reply:
The libraries aren't specifically designed with this use case in mind, they **can** be combined like we do here, but they all have their own separate use case.

* crownstone_cloud: uses polling to fetch all user data from the cloud, and also update it's own data structure (including the state)
* crownstone_sse: Listens to server send events that are send upon update in cloud data, which **can** then be used to update the state instead of having crownstone_cloud use polling to get the state;
* crownstone_uart: set up a connection with an usb dongle and listen to events through it's own eventbus, to provide state updates and other data

Neither crownstone_see and crownstone_uart are aware of any user data or provide a data structure. They just provide updates, and that's it. 

If we wanted to abstract the state updates away from the user, then we would have to setup listeners in crownstone_cloud, and then it would **depend** on the other 2 libraries, which we don't want. If the user wants to update their data structure (can be the one in crownstone_cloud, or even something completely different) they can setup listeners to receive state updates and then they have to update the attributes in their own data, like we do here.